modmail/dispatcher.py

```diff
@@ -0,0 +1,167 @@
+import asyncio
+import bisect
+import logging
+from typing import Callable, Coroutine, Dict, List, Optional, Union
+
+from modmail import ModmailLogger
+
+logger: ModmailLogger = logging.getLogger(__name__)
+
+CoroutineFunction = Callable[..., Coroutine]
+
+
+class Dispatcher:
+    """
+    Dispatches events through an async event handler system.
+
+    Supports blocking events and priority.
+    """
+
+    # These are separate because it makes using the bisect module easier. See _register_handler.
+    blocking_handlers: Dict[str, List[Optional[CoroutineFunction]]]
+    blocking_priorities: Dict[str, List[int]]
```

Comment:
What are the supported priorities? How does this handle priorities between blocking_priorities and blocking_handlers. Some handlers may have a high priority, but aren't blocking.

Reply:
Any integer priority level is supported. I suppose any float is also supported, but not explicitly. 

Handlers with a high priority but not "blocking" must necessarily be blocking as a group at minimum to prevent further events after them from being called too soon. However, because its much more complex to implement per-step blocking as opposed to just linear, I figured it's just much easier for everyone if they want a high priority handler to support the ability to block others. 

The only reason non-prioritized handlers don't support event cancellation is because they're all fired at once.

Reply:
I've added a nonblocking decorator to general utils if someone wants to explicitly opt into that behavior